### DISCUSSION TITLE
Feature Request: support join elimination rule

### DISCUSSION THREAD

**Discussion by iwanttobepowerful** [duckdb/duckdb#6464] [Ideas]
reference:
[blog.jooq.org/2017/09/01/join-elimination-an-essential-optimiser-feature-for-advanced-sql-usage](https://blog.jooq.org/2017/09/01/join-elimination-an-essential-optimiser-feature-for-advanced-sql-usage/)

and this paper《Outer Join Elimination in the Teradata RDBMS》

**Comment by Alex-Monahan** [comment]
Anything you can provide as to why you are interested in this feature and how it could help you?

**Reply by iwanttobepowerful** [reply]
Query rewrite plays an important role in optimizing complex and tool generated queries. One important such technique is eliminating redundant joins whenever possible.

**Comment by Reminiscent** [comment]
I am very interested about this feature. Can I try to design and develop this one? @Mytherin

**Reply by iwanttobepowerful** [reply]
nice! maybe you can ref your tidb's solution

**Reply by Mytherin** [reply]
Sure, do you have any initial ideas for a design that we could discuss?

**Comment by Reminiscent** [comment]
The basic idea is that for outer joins, each row of the outer side will definitely appear once or multiple times in the join result set. When the row of the outer side cannot find a match or can only find one matching row, this row of the outer side will only appear once in the join result; when the row of the outer side can find multiple matching rows, it will appear multiple times in the join result. Therefore, if the inner side satisfies the uniqueness property on the join key, it is impossible for the rows of the outer side to find multiple matching rows, so each row of the outer side will only appear once in the join result. At the same time, if the upper-level(father) operator only needs the data of the outer side, the outer join can be directly eliminated from the query. Similarly, it is easy to understand that when the upper-level(father) operator only needs the deduplicated result of the outer side, the outer join can also be eliminated.

So the elimination of outer join requires certain conditions to be met:
1. The parent operator of LogicalJoin only uses the columns output by the outer side of LogicalJoin.
2. The output columns for the outer part are uniqueness:
   1. The join key in LogicalJoin satisfies uniqueness property in the output of the inner plan.
   2. The parent operator of LogicalJoin deduplicates the input records.

Both Condition 1 and Condition 2 must be met simultaneously, but only one of Condition 2.1 and Condition 2.2 needs to be satisfied.

Case1: Meeting Condition 1 and Condition 2.1
"select t1.a from t1 left join t2 on t1.b = t2.pk;" -> "select t1.a from t1;"

Case2: Meeting Condition 1 and Condition 2.1
"select distinct(t1.a) from t1 left join t2 on t1.b = t2.b;" -> "select distinct(t1.a) from t1;"

Moreover, there are additional details that need to be considered, and we will start with the simpler scenarios and expand the scope step by step.

The above is the principle for outer join elimation optimization. For more details:
1. The rule will be put behind the 'filter pushdown'.
2. Traverse the plan tree from top to bottom and perform different operations based on the different operators.
3. For non-join operators, we need to collect the columns used by the operator and pass them down layer by layer. This is used for the determination of condition 1.
4. For aggregation operators, we need to collect the columns which is used in distinct clause and pass them down layer by layer. This is used for the determination of condition 2.2.
5. For join operator:
   1. We need to verify whether the columns used by the parent operator not belong to the inner side. -> condition 1
   2. We need to check if the rows outputted by the outer side can be de-duplicated by the upper layer's DISTINCT which is been record before. -> condition 2.2
   3. We need to check whether the join keys for the inner side are uniqueness, for example: unique index or unique column. -> condition 2.1

@Mytherin Can you help me to review the design?

**Reply by iwanttobepowerful** [reply]
```sql
create table t1 (a int);
create table t2 (a int, b int);
create table t3 (a int, b int);

select t1.* from t1 left join
(
  select t2.b as v2b, count(*) as v2c
  from t2 left join t3 on t3.a=t2.b
  group by t2.b
) v2
on v2.v2b=t1.a;

select distinct t1.a from t1 left join t2 on t1.a = t2.a;

select max(t1.a) from t1 left join t2 on t1.a = t2.a;
select min(t1.a) from t1 left join t2 on t1.a = t2.a;
```

**Reply by Reminiscent** [reply]
Thank you for your input. While there are several details to consider, we can start by addressing the normal case and gradually expanding the scope. Here are some of the key issues we need to consider:

1. Does the unique index contain multiple columns? For each individual table on the inner side, we need to ensure that all columns involved in its join conditions are a superset of some unique index, which can guarantee the output of only one row.
2. Does the inner table contain multiple tables? To eliminate the entire LEFT JOIN, all tables on the inner side must ensure a unique output of one row.
3. Does the inner table include a new LEFT JOIN? We need to perform a recursive traversal to determine whether the inner LEFT JOIN can be eliminated. After elimination, we can examine the outer layer.
4. Is the inner table itself a derived table (subquery)? The subquery must contain a group by clause, and the group by columns must be the primary key columns of the derived table from the external perspective.
5. Which operation in the parent node can de-duplicate the output of the outer side? This may include certain aggregation functions or other operations.

**Reply by iwanttobepowerful** [reply]
@Mytherin Hi, Mark, could you please review this design and give @Reminiscent any adcvices?

**Reply by Mytherin** [reply]
> The basic idea is that for outer joins, each row of the outer side will definitely appear once or multiple times in the join result set. When the row of the outer side cannot find a match or can only find one matching row, this row of the outer side will only appear once in the join result; when the row of the outer side can find multiple matching rows, it will appear multiple times in the join result. Therefore, if the inner side satisfies the uniqueness property on the join key, it is impossible for the rows of the outer side to find multiple matching rows, so each row of the outer side will only appear once in the join result. At the same time, if the upper-level(father) operator only needs the data of the outer side, the outer join can be directly eliminated from the query. Similarly, it is easy to understand that when the upper-level(father) operator only needs the deduplicated result of the outer side, the outer join can also be eliminated.
> 
> So the elimination of outer join requires certain conditions to be met:
> 
> 1. The parent operator of LogicalJoin only uses the columns output by the outer side of LogicalJoin.
> 2. The output columns for the outer part are uniqueness:
>    
>    1. The join key in LogicalJoin satisfies uniqueness property in the output of the inner plan.
>    2. The parent operator of LogicalJoin deduplicates the input records.
> 
> Both Condition 1 and Condition 2 must be met simultaneously, but only one of Condition 2.1 and Condition 2.2 needs to be satisfied.
> 
> Case1: Meeting Condition 1 and Condition 2.1 "select t1.a from t1 left join t2 on t1.b = t2.pk;" -> "select t1.a from t1;"
> 
> Case2: Meeting Condition 1 and Condition 2.1 "select distinct(t1.a) from t1 left join t2 on t1.b = t2.b;" -> "select distinct(t1.a) from t1;"
> 
> Moreover, there are additional details that need to be considered, and we will start with the simpler scenarios and expand the scope step by step.
> 
> The above is the principle for outer join elimation optimization. For more details:
> 
> 1. The rule will be put behind the 'filter pushdown'.
> 2. Traverse the plan tree from top to bottom and perform different operations based on the different operators.
> 3. For non-join operators, we need to collect the columns used by the operator and pass them down layer by layer. This is used for the determination of condition 1.
> 4. For aggregation operators, we need to collect the columns which is used in distinct clause and pass them down layer by layer. This is used for the determination of condition 2.2.
> 5. For join operator:
>    
>    1. We need to verify whether the columns used by the parent operator not belong to the inner side. -> condition 1
>    2. We need to check if the rows outputted by the outer side can be de-duplicated by the upper layer's DISTINCT which is been record before. -> condition 2.2
>    3. We need to check whether the join keys for the inner side are uniqueness, for example: unique index or unique column. -> condition 2.1
> 
> @Mytherin Can you help me to review the design?

That sounds like a good plan to me. I would try to do this in steps and focus on either 2.1 or 2.2 first, and then expand the scope of the optimization later on.

Note that it is also possible to eliminate INNER joins, as long as you can guarantee that every row will find exactly one match. This is the case if there is a join between a primary-foreign key pair without filters.

**Comment by djouallah** [comment]
snowflake implemented this but they use relationship hint

https://hoffa.medium.com/how-to-help-snowflake-eliminate-your-redundant-joins-with-constraints-f6c696264114

**Reply by Reminiscent** [reply]
Thank you for your input. I'll take a look.

**Comment by iwanttobepowerful** [comment]
Hi, @Mytherin When will this feature be implemented? ^_^

**Reply by Reminiscent** [reply]
Sorry. I was busy before. I am starting to implement this feature. Maybe I can submit the simpler version of PR in a few days.

**Reply by iwanttobepowerful** [reply]
thx~

**Reply by Reminiscent** [reply]
https://github.com/duckdb/duckdb/pull/7668 Hi @iwanttobepowerful I have implemented the simpler version. I'll gradually expand the scope after this PR has been merged.

**Reply by iwanttobepowerful** [reply]
nice!

**Comment by DanCory** [comment]
It doesn't look like this was ever implemented. Query23 in TPCDS is a candidate for join elimination, if foreign keys are declared. It has an unnecessary join with the customer table which can be replaced by just checking the foreign key is not null. At scale factor 10, the query completes on my machine in about 14s with the customer table and 12s without.

Rewritten query23 is attached 
[query23.zip](https://github.com/user-attachments/files/20178301/query23.zip)

**Comment by exaroj** [comment]
Join elimination is useful for BI dashboarding where you have a complex schema and you don't know what filters or groupings are going to be applied. (As @iwanttobepowerful mentioned.)

If you have effective join elimination you can have a single query/view with all the dimensions present and expect them to be removed if unreferenced. If you don't then you need to dynamically generate SQL for each use case.

Other tools have `RELY` / `RELIABLE`. dbt et al have cardinality qualifiers.

[src/optimizer/join_elimination.cpp](https://github.com/duckdb/duckdb/blob/985b2f2515adf44ea65dd4239c7bfb30981b94dc/src/optimizer/join_elimination.cpp) does some optimization but I think this can never achieve what is needed: you can not generally infer the cardinality of the RHS except in a few special cases (DISTINCT, tables with constraints, columns with precomputed statistics).

It would be really useful to have explicit syntax for `table1 MANY TO ONE JOIN table2` (cf SAP) to enable join elimination and optimization. However, I can see this is a [footgun](https://en.wiktionary.org/wiki/footgun) for many users.

https://docs.snowflake.com/en/user-guide/join-elimination
https://community.sap.com/t5/technology-blog-posts-by-sap/introduction-of-join-cardinalities-in-sql-with-sap-hana-2-0-sps04/ba-p/13398965
